dev/ledger.md

```diff
@@ -1030,13 +1040,15 @@ point must be discarded and the entire transaction rejected.
 
 ### Application Stateful TEAL Execution Semantics
 
-- During the execution of an `ApprovalProgram` or `ClearStateProgram`, the
-  application’s `LocalStateSchema` and `GlobalStateSchema` may never be
-  violated. The program's execution will fail on the first instruction that
-  would cause the relevant schema to be violated. Writing a `Bytes` value to a
-  local or global [TEAL Key/Value Store][TEAL Key/Value Stores] longer than 64
-  bytes, or writing any value to a key longer than 64 bytes, will likewise cause
-  the program to fail on the offending instruction.
+- During the execution of an `ApprovalProgram` or `ClearStateProgram`,
+  the application’s `LocalStateSchema` and `GlobalStateSchema` may
+  never be violated. The program's execution will fail on the first
+  instruction that would cause the relevant schema to be
+  violated. Writing a `Bytes` value to a local or global [TEAL
+  Key/Value Store][TEAL Key/Value Stores] such that the sum of the
+  lengths of the key and value in bytes exceeds 128, or writing any
```

Comment:
Can applications with TEAL v2 or v3 benefit from the new rule that key size + value size must not exceed 128 bytes instead of key size and value size individually must not exceed 64 bytes?

Reply:
Yes.  I suppose there is, again, the theoretical issue that an old smart contract could have been *intending* to panic if its value exceeded 64.

Reply:
I don't follow. Does that mean that stateful TEAL programs are not versioned? So a stateful TEAL v2 program that would be called after the upgrade will run according to TEALv4 rules?

Reply:
TEAL programs are versioned, but these things are done outside of the TEAL environment.  TEAL doesn't force values to be less than 64 bytes, the rules for writing ApplyData does.  Similarly, TEAL doesn't set program sizes, the transaction processing rules do.

Reply:
Yes, I think it's fine. In theory, there is a security issue.
In practice, it is extremely unlikely someone did that.